### DISCUSSION TITLE
Need help finding the entry point to modify a table.

### DISCUSSION THREAD

**Discussion by DiagProf** [GGG-KILLER/Loretta#123] [Q&A] [answered]
Hello, in my hobby project 'EcuDiagSim,' I aim to modify some Lua files. Initially, I considered using Regex for this task, but then I discovered this project, which seems much more suited for my needs. However, I've realized I lack the knowledge on how to get started, particularly in finding and modifying a specific table named 'Raw' within these files. Could you provide idea on how to use Loretta to locate and manipulate the 'Raw' table in a Lua file? In my Lua scripts, the 'Raw' table is always a first-level sub-element of a top-level table. I hope I've explained that correctly. Here's an example of how the Lua code should look before and after the modifications. 
before rework:
```lua
RandomName = {
	RequestId = 0x7A6,
	ResponseId = 0x7A7,
	Raw = {
		["10 03"] = "50 03 00 28 00 C8",
		["3E 00"] = "7E 00",
		["22 22 35"] = "62 22 35 47 11",
		["2E 22 35 47 11"] = "6E 22 35",
               -- Additional entries here
    }
}
```

Aftter rework. This is what the result should look like.
```lua
did = {
	["22 35"] = "47 11",
}
RandomName = {
	RequestId = 0x7A6,
	ResponseId = 0x7A7,
	Raw = {
		["10 03"] = "50 03 00 28 00 C8",
		["3E 00"] = "7E 00",
		["22 22 35"] = function(r) return "62 22 35 " .. did[request:sub(4)] end,
		["2E 22 35 *"] = function(request) did[request:sub(4,8)] = request:sub(10) return "6E 22 35" end,
                -- Additional entries here
    }
}
```
Please don't get me wrong; I'm not looking for anyone to do my work for me, but I'd really appreciate some tips. I've experimented with <TableConstructorExpressionSyntax> but I'm not sure if I'm on the right track. Any advice or pointers would be greatly appreciated.

Thank you in advance for your assistance.

**Comment by GGG-KILLER** [comment, marked as answer]
Hello, thank you for giving our project a shot!

For this you could either do a whole-program search through the tree's root for a table has field named `Raw` or do it all at once with a `LuaSyntaxRewriter`. For the sake of brevity I'm gonna demonstrate only the whole-program search but if you become more interested in the `LuaSyntaxRewriter` approach, I can try to guide you on it as well.

For example:
```cs
using Loretta.CodeAnalysis;
using Loretta.CodeAnalysis.Lua;
using Loretta.CodeAnalysis.Lua.Syntax;

var code = """
RandomName = {
    RequestId = 0x7A6,
    ResponseId = 0x7A7,
    Raw = {
        ["10 03"] = "50 03 00 28 00 C8",
        ["3E 00"] = "7E 00",
        ["22 22 35"] = "62 22 35 47 11",
        ["2E 22 35 47 11"] = "6E 22 35",
        -- Additional entries here
    }
}
""";

var tree = LuaSyntaxTree.ParseText(code, new LuaParseOptions(LuaSyntaxOptions.Lua51));
var root = tree.GetCompilationUnitRoot();

var tables = root.DescendantNodes().OfType<TableConstructorExpressionSyntax>();
var toAnalyze = new List<TableConstructorExpressionSyntax>();

foreach (var table in tables)
{
    // This checks that the table field is in the form Raw = { ... }
    if (table.Fields.Any(x => x is IdentifierKeyedTableFieldSyntax { Identifier.Value: "Raw", Value: TableConstructorExpressionSyntax }))
    {
        toAnalyze.Add(table);
    }
    // If you want to handle ["Raw"] = { ... } case as well
    else if (table.Fields.Any(x => x is ExpressionKeyedTableFieldSyntax
    {
        Key: LiteralExpressionSyntax { RawKind: (int)SyntaxKind.StringLiteralExpression, Token.Value: "Raw" },
        Value: TableConstructorExpressionSyntax
    }))
    {
        toAnalyze.Add(table);
    }
}
```

and then, after you have listed the tables you want to modify, you can then use the [`SyntaxNode.ReplaceNodes()`](https://loretta-oss.dev/api/Loretta.CodeAnalysis.SyntaxNodeExtensions#Loretta_CodeAnalysis_SyntaxNodeExtensions_ReplaceNodes__2___0_IEnumerable___1__Func___1___1_Loretta_CodeAnalysis_SyntaxNode__) function to replace all the tables.

But since you also want to add a local table to hold some state before the original one, you'll actually need to work on the `StatementList` that the table is located on, which you can do by using `.FirstAncestorOrSelf<StatementListSyntax>()` on the table expression.
Once you have the statement list, you can then insert the local which will hold the state before the statement that holds the `Raw` table and then modify the statement that contains the raw table.

To this end, you can either use a `LuaSyntaxRewriter` which would be more involved, or you can group the tables by their containing statement list and then call `.ReplaceNodes` on the statement lists and doing the work of inserting the locals and rewriting the statements containing the raw table.

Hope this helps!

**Reply by DiagProf** [reply]
Thank you for the quick response! It's the weekend now, so I can make some more progress on my project. I understand how to find a table containing a 'Raw' table. However, I'm still figuring out how to navigate the syntax tree to work exclusively with the 'Raw' table for my analysis. Could you provide an example on isolating the 'Raw' table within the syntax tree? I believe I still lack a lot of knowledge. Could you explain the approach with the LuaSyntaxRewriter too? But only if your time allows. 
I would appreciate any help.

**Reply by GGG-KILLER** [reply]
The `Raw` table isolation is already being done in the code I sent, the `toAnalyze` has only the `Raw` tables themselves.

And about the LuaSyntaxRewriter, I believe the tutorial is a good reference to start off with: https://loretta-oss.dev/tutorials/creating-a-localizer/part-1?tabs=programcs-1%2Cprogramcs-2%2Cprogramcs-3

**Reply by DiagProf** [reply]
Thank you for your tips, and apologies for my late response; I only get to this on weekends. I've now uploaded my work to GitHubRepo https://github.com/DiagProf/EcuDiagSimLuaDynamization  . Could you assist me further... what would be the correct approach to access the name of the table that contains the 'Raw' table when I'm within the 'Raw' table?  In this case, it's 'RandomName'

i find a solution. Not sure is the best but works.
```csharp
                        //Test do get the EcuName (outer Table name) starting from the Raw Table
                        string ecuName = "notFoundEcuName";
                        var parentTable = tableConstructorExpressionSyntax.Ancestors().OfType<TableConstructorExpressionSyntax>().FirstOrDefault()?.Parent?.Parent;
                        if (parentTable.IsKind(SyntaxKind.AssignmentStatement))
                        {
                            var firstVariable = ((parentTable as AssignmentStatementSyntax)!).Variables.FirstOrDefault();
                            if ( firstVariable.IsKind(SyntaxKind.IdentifierName) )
                            {
                               
                                ecuName = ((firstVariable as IdentifierNameSyntax)!).Name;
                            }
                        }
                        Console.WriteLine($"{ecuName}");
 ```

**Reply by DiagProf** [reply]
Sorry, I have another question. I'm now at the point where I want to write new code, but I'm not sure how. Could you please take a look?
```csharp
    class WriteDidRewriter : LuaSyntaxRewriter
    {
        ...
        public override SyntaxNode? VisitExpressionKeyedTableField(ExpressionKeyedTableFieldSyntax node)
        {
            // Input 
            //e.g.ExpressionKeyedTableFieldSyntax ExpressionKeyedTableField ["2E 22 35 47 11"] = "6E 22 35",

            //ToDo rework  e.g. -> ["2E 22 35 *"] = function(request) return updateDidData(YourEcuName,request ) end,  
            //node.Update();

            return base.VisitExpressionKeyedTableField(node);
        }
    }
```
Thanks for your help.

**Reply by GGG-KILLER** [reply]
Hello, I apologize for the delay in replying, I've been very busy with work lately.

> i find a solution. Not sure is the best but works.

Yeah, it is a very acceptable solution, you're traversing the tree which is the standard way to do so

> Sorry, I have another question. I'm now at the point where I want to write new code, but I'm not sure how. Could you please take a look?

In that case, you could do the following to create a function:
```cs
var function = SyntaxFactory.AnonymousFunctionExpression(
    SyntaxFactory.ParameterList(
        SyntaxFactory.SeparatedList<ParameterSyntax>(new[]{ SyntaxFactory.NamedParameter("request") })
    ),
    SyntaxFactory.StatementList(new[]
    {
        SyntaxFactory.ReturnStatement(
            // updateDidData(YourEcuName, request)
            SyntaxFactory.FunctionCall(
                SyntaxFactory.IdentifierName("updateDidData"),
                SyntaxFactory.ExpressionListFunctionArgument(
                     SyntaxFactory.SeparatedList<ExpressionSyntax>(new[]
                     {
                         SyntaxFactory.IdentifierName("YourEcuName"),
                         SyntaxFactory.IdentifierName("request")
                     })
                )
            )
        )
    })
);
```

and then update the node in your rewriter by doing the following:
```cs
    class WriteDidRewriter : LuaSyntaxRewriter
    {
        ...
        public override SyntaxNode? VisitExpressionKeyedTableField(ExpressionKeyedTableFieldSyntax node)
        {
            var key = // your logic for editing the key and replacing "2E 22 35 47 11" with "2E 22 35 *", this should be an ExpressionSyntax
            var function = // the code above
            return base.WithKey(base).WithValue(function).NormalizeWhitespace();
        }
    }
```
